control_plane/src/background_process.rs

```diff
@@ -98,20 +98,34 @@ where
         InitialPidFile::Expect(path) => path,
     };
 
-    let mut spawned_process = filled_cmd.spawn().with_context(|| {
+    let spawned_process = filled_cmd.spawn().with_context(|| {
         format!("Could not spawn {process_name}, see console output and log files for details.")
     })?;
     let pid = spawned_process.id();
     let pid = Pid::from_raw(
         i32::try_from(pid)
             .with_context(|| format!("Subprocess {process_name} has invalid pid {pid}"))?,
     );
+    // set up a scopeguard to kill & wait for the child in case we panic or bail below
+    let spawned_process = scopeguard::guard(spawned_process, |mut spawned_process| {
+        println!("SIGKILL & wait the started process");
+        (|| {
+            // TODO: use another signal that can be caught by the child so it can clean up any children it spawned (e..g, walredo).
+            spawned_process.kill().context("SIGKILL child")?;
+            spawned_process.wait().context("wait() for child process")?;
+            anyhow::Ok(())
+        })()
+        .with_context(|| format!("scopeguard kill&wait child {process_name:?}"))
+        .unwrap();
+    });
 
     for retries in 0..RETRIES {
         match process_started(pid, pid_file_to_check, &process_status_check).await {
             Ok(true) => {
-                println!("\n{process_name} started, pid: {pid}");
-                return Ok(spawned_process);
+                println!("\n{process_name} started and passed status check, pid: {pid}");
+                // leak the child process, it'll outlive this neon_local invocation
```

Comment:
If you do ctrl+c for the neon_local process, does it also kill the pageserver? It seems no.

Reply:
Would be sort of easy to take SIGINT as cancellation, but unsure if it's really necessary given what we use `neon_local` for.

Reply:
> If you do ctrl+c for the neon_local process, does it also kill the pageserver? It seems no.

1. pre-existing issue, not a reason to block this PR
2. I think process group forwarding will forward the SIGINT to the pageserver.
3. The default SIGINT action is to terminate the process.
4. Pageserver only registers for special SIGINT at the end of its startup.
5. I think if it were to start any walredo processes, these would also receive the SIGINT, again, due to process group signal forwarding.
6

Reply:
> pre-existing issue, not a reason to block this PR

Fair point, but it feels weird to me that a PR with a description "After this PR, there is no leaked process." adds a line "leak the child process" to the code, and removes `Child` from the return type of a lot of functions. But you are right, we can always improve it later, so let's do it later.

Reply:
That added line about "leak the child process" is a different kind of leakage.

Reply:
Yeah, you are right, it's just confusing. But fine to merge in current state, so resolving.